Fix tests requiring timezone alignment

bpintea opened the pull request:
This PR will have the test comparing H2 against ES results have the
H2 timezone sent to ES. This ensures that some timezone sensitive tests
whose results otherwise differ, will always work, no matter the
randomised timezone chosen for the test.

Closes #37045

bpintea committed:
Fix tests requiring timezone alignment

This commit will have the test comparing H2 against ES results have the
H2 timezone sent to ES. This ensures that some timezone sensitive tests
whose results otherwise differ, will always work, no matter the
randomized timezone chosen for the test.

elasticmachine commented:
Pinging @elastic/es-search (:Search/SQL)

bpintea committed:
remove disallowed API

- removing use of String.toUpperCase().

matriv commented on x-pack/plugin/sql/qa/src/main/resources/datetime.sql-spec:
```diff
@@ -119,8 +119,6 @@ dayOfWeekGroupBy
 SELECT DAY_OF_WEEK(birth_date) day, COUNT(*) c FROM test_emp WHERE DAY_OF_WEEK(birth_date) < 6 GROUP BY day ORDER BY DAY_OF_WEEK(birth_date);
 
 // current
```
probably move/remove this comment as well?

matriv replied:
Also move/remove this comment?

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

matriv reviewed (commented):
I think this solution doesn't solve the whole issue.
On the change of the minute/hour/day/month/year it can be that one test (H2) is run before the change of the corresponding field and the ES test after the change.

matriv reviewed (commented):
Since this is definitely an improvement, I'd say to proceed and merge it first in master and see how the CI runs will behave. The one that I feel could be flaky is the minute one.

Could you please also add tests for CURRENT_DATE/CURRENT_TIME?

costin commented:
If using H2 is problematic how about doing this test programmatically?

That said, I'm still confused on why, if this fix is needed, it affects only some datetime tests and not all of them? If the TZ is different, so should be the results.

bpintea committed:
Prevent TZ-tests from running before hour change

Prevent tests requring TZ sync'ing from running before the hour changes
(which could trigger a day/month/year change).

Also comment better why TZ-sync'ing is required.

bpintea committed:
Add more CURRENT_DATE/_TIME tests

Also change the MINUTE() tests to work with static values; the dynamic
"current" ones can change between runs against H2 and ES.

bpintea commented:
> Since this is definitely an improvement, I'd say to proceed and merge it first in master and see how the CI runs will behave. The one that I feel could be flaky is the minute one.

Sure. I've also added a guard against running before hour change.

> 
> Could you please also add tests for CURRENT_DATE/CURRENT_TIME?

Added those too.

bpintea commented:
> That said, I'm still confused on why, if this fix is needed, it affects only some datetime tests and not all of them? If the TZ is different, so should be the results.

I've added some more comments to explain why this happens:
https://github.com/elastic/elasticsearch/blob/12784ee92f3a223e620e050bfc381fd9e85d0ce3/x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/SpecBaseIntegrationTestCase.java#L123

But not all tests require that:
https://github.com/elastic/elasticsearch/blob/12784ee92f3a223e620e050bfc381fd9e85d0ce3/x-pack/plugin/sql/qa/src/main/resources/datetime.sql-spec#L133

bpintea committed:
fix code style not cought by prechecks

s/final private/private final/

bpintea committed:
fix API usage with defaults

LocalDateTime.now() needs to be given a ZoneId

bpintea committed:
Further fixes for tests

- MINUTE requires TZ sync when extracting from a timestamp, but needs
muting when comes from a time (H2 not having timezones attached to
those)

bpintea commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'master' into fix/timezone_sensitive_tests

astefan commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/SqlSpecTestCase.java:
```diff
@@ -13,15 +13,19 @@
  * with those executed against H2's jdbc client.
  */
 public abstract class SqlSpecTestCase extends SpecBaseIntegrationTestCase {
+    private final int TZSYNC_MINUTE_THREASHOLD = 55;
```
The name of the constant, I believe, has a typo in it: THREASHOLD vs THRESHOLD.

bpintea replied:
thanks, fixed it.

astefan reviewed (approved):
LGTM. Left one comment.

matriv reviewed (approved):
LGTM

bpintea committed:
Address review comments

s/THREASHOLD/THRESHOLD/
